openwpm/deploy_browsers/deploy_firefox.py

```diff
@@ -110,16 +117,30 @@ def deploy_firefox(
         # TODO restore detailed logging
         # fo.set_preference("extensions.@openwpm.sdk.console.logLevel", "all")
 
+    # Geckodriver currently places the user.js file in the wrong profile
+    # directory, so we have to create it manually here.
+    # TODO: Remove this workaround once
+    # https://github.com/mozilla/geckodriver/issues/1844 is fixed.
+    # Load existing preferences from the profile's user.js file
+    prefs = configure_firefox.load_existing_prefs(browser_profile_path)
+    # Load default geckodriver preferences
+    prefs.update(configure_firefox.DEFAULT_GECKODRIVER_PREFS)
+    # Pick an available port for Marionette (https://stackoverflow.com/a/2838309)
+    s = socket.socket()
+    s.bind(("", 0))
+    marionette_port = s.getsockname()[1]
+    s.close()
+    prefs["marionette.port"] = marionette_port
```

Comment:
> The way I've implemented picking a port and passing it to geckodriver has a race condition, as another process may get the port before Marionette does. I think this will rarely happen though. Is it something we can accept? Can we think of a better way to do this?

This seems acceptable. I agree with Stefan that a comment noting this is possible will be helpful.

What happens if this fails? I assume we'll automatically retry?

Reply:
Yes, the browser manager is automatically restarted